comms/src/peer_manager/peer_storage.rs

```diff
@@ -0,0 +1,682 @@
+            self.net_address_hm
+                .insert(net_address_with_stats.net_address.clone(), index);
+        }
+    }
+
+    /// Add a single peer to the datastore using the provided index as a key
+    fn add_peer_to_datastore(&mut self, index: usize, peer: &Peer<PubKey>) -> Result<(), PeerManagerError> {
+        if let Some(ref mut datastore) = self.datastore {
+            let index_bytes = index.to_binary().map_err(|e| PeerManagerError::SerializationError(e))?;
+            let peer_bytes = peer.to_binary().map_err(|e| PeerManagerError::SerializationError(e))?;
+            datastore
+                .put_raw(&index_bytes, peer_bytes)
+                .map_err(|e| PeerManagerError::DatastoreError(e))?;
+        }
+        Ok(())
+    }
+
+    /// Remove the peer specified by a given index from the datastore and remove hashmap keys
+    fn remove_from_db_and_links(&mut self, index: usize) -> Result<(), PeerManagerError> {
+        let peer_count = self.peers.len();
+        if (index > 0) && (index < peer_count) {
+            // Remove entry from datastore
+            if let Some(ref mut datastore) = self.datastore {
+                let index_bytes = index.to_binary().map_err(|e| PeerManagerError::SerializationError(e))?;
+                datastore
+                    .delete_raw(&index_bytes)
+                    .map_err(|e| PeerManagerError::DatastoreError(e))?;
+            }
+            // Remove hashmap links
+            self.node_id_hm.remove(&self.peers[index].node_id);
+            self.public_key_hm.remove(&self.peers[index].public_key);
+            for net_address_with_stats in &self.peers[index].addresses.addresses {
+                self.net_address_hm.remove(&net_address_with_stats.net_address);
+            }
+            Ok(())
+        } else {
+            Err(PeerManagerError::IndexOutOfBounds)
+        }
+    }
+
+    /// Retrieve a single peer from the data store using the provided index
+    pub fn get_peer_from_datastore(&self, index: usize) -> Result<Peer<PubKey>, PeerManagerError> {
+        match self.datastore {
+            Some(ref datastore) => {
+                let index_bytes = index.to_binary().map_err(|e| PeerManagerError::SerializationError(e))?;
+                match datastore
+                    .get_raw(&index_bytes)
+                    .map_err(|e| PeerManagerError::DatastoreError(e))?
+                {
+                    Some(peer_raw) => Peer::<PubKey>::from_binary(peer_raw.as_slice())
+                        .map_err(|_| PeerManagerError::DeserializationError),
+                    None => Err(PeerManagerError::EmptyDatastoreQuery),
+                }
+            },
+            None => Err(PeerManagerError::DatastoreUndefined),
+        }
+    }
+
+    /// Enables Thread safe access - Changes the ban flag bit of the peer
+    pub fn set_banned(&mut self, node_id: &NodeId, ban_flag: bool) -> Result<(), PeerManagerError> {
```

Comment:
This says its thread safe in the comment but I don't see any thread safety stuff? Has it been moved to another part of the system?